crates/cow-amm/src/maintainers.rs

```diff
@@ -0,0 +1,79 @@
+use {
+    crate::{cache::Storage, Amm},
+    contracts::ERC20,
+    ethcontract::futures::future::join_all,
+    ethrpc::Web3,
+    shared::maintenance::Maintaining,
+    std::sync::Arc,
+    tokio::sync::RwLock,
+};
+
+pub struct EmptyPoolRemoval {
+    storage: Arc<RwLock<Vec<Storage>>>,
+    web3: Web3,
+}
+
+impl EmptyPoolRemoval {
+    pub fn new(storage: Arc<RwLock<Vec<Storage>>>, web3: Web3) -> Self {
+        Self { storage, web3 }
+    }
+
+    /// Checks if the given AMM has a zero token balance.
+    async fn check_single(&self, amm: Arc<Amm>) -> bool {
+        let amm_address = amm.address();
+        let futures = amm.traded_tokens().iter().map(move |token| async move {
+            match ERC20::at(&self.web3, *token)
+                .balance_of(*amm_address)
+                .call()
+                .await
+            {
+                Ok(balance) => balance.is_zero(),
+                Err(err) => {
+                    tracing::warn!(
+                        amm = ?amm_address,
+                        ?token,
+                        ?err,
+                        "failed to check AMM token balance"
+                    );
+                    false
+                }
+            }
+        });
+        join_all(futures).await.into_iter().any(|is_empty| is_empty)
```

Comment:
nit: could use `select_ok` or the like to return as soon as one balance check is 0 (no need to wait for all futures to complete), but probably no big performance gain in practice.

Reply:
Oh, nice, thanks. I was looking for something like that. Implemented, the reversed logic looks a bit confusing, though.